presto-main/src/main/java/com/facebook/presto/execution/scheduler/FixedSourcePartitionedScheduler.java

```diff
@@ -165,21 +182,30 @@ public ScheduleResult schedule()
         // schedule a task on every node in the distribution
         List<RemoteTask> newTasks = ImmutableList.of();
         if (!scheduledTasks) {
-            OptionalInt totalPartitions = OptionalInt.of(nodes.size());
             newTasks = Streams.mapWithIndex(
                     nodes.stream(),
-                    (node, id) -> stage.scheduleTask(node, toIntExact(id), totalPartitions))
+                    (node, id) -> stage.scheduleTask(node, toIntExact(id)))
                     .filter(Optional::isPresent)
                     .map(Optional::get)
                     .collect(toImmutableList());
             scheduledTasks = true;
+
+            // notify listeners that we have scheduled all tasks so they can set no more buffers or exchange splits
+            stage.transitionToFinishedTaskScheduling();
         }
 
         boolean allBlocked = true;
         List<ListenableFuture<?>> blocked = new ArrayList<>();
         BlockedReason blockedReason = BlockedReason.NO_ACTIVE_DRIVER_GROUP;
 
         if (groupedLifespanScheduler.isPresent()) {
+            while (!tasksToRecover.isEmpty()) {
+                if (anySourceSchedulingFinished) {
+                    throw new IllegalStateException("Recover after any source scheduling finished is not supported");
+                }
+                groupedLifespanScheduler.get().onTaskFailed(tasksToRecover.poll(), sourceSchedulers);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentLinkedQueue`, but your usage of `isEmpty()` and `poll()` may not be thread-safe at lines: **202 and 206**. If the `ConcurrentLinkedQueue` contains only **one element** and two threads perform this **same check at the same time** , the second thread&#39;s call to `poll()` will get `null`. This `null` may be used at line: **206**.

**Fix**  
 Consider calling `poll()`, checking instead of your current check if the returned object is `null`, and only then using that object.